### PR TITLE
Improve type resolution

### LINKED ISSUE #32044: Compiler crashes with record type inclusion with overriding field
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public type Foo record {
    string code?;
    string message?;
};

public type Bar record {
    string processId?;
    Qux[] reasons?;
    boolean success?;
};

public type Baz record {
    *Bar;
    string id?;
    Foo[] reasons?;
};

public type Qux record {
    string code?;
    string message?;
};
```

```log
[2021-08-06 16:27:20,623] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameType(Types.java:341)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameType(Types.java:306)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:636)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.checkRecordEquivalency(Types.java:1564)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.checkStructEquivalency(Types.java:1471)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:785)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isArrayTypesAssignable(Types.java:1152)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:805)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isAssignable(Types.java:631)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.lambda$resolveIncludedFields$33(SymbolEnter.java:4480)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.resolveIncludedFields(SymbolEnter.java:4492)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.resolveFields(SymbolEnter.java:3614)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineFieldsOfObjectOrRecordTypeDef(SymbolEnter.java:3552)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineFields(SymbolEnter.java:3506)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineConstructs(SymbolEnter.java:404)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:318)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:165)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:275)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.definePackage(SymbolEnter.java:268)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:377)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:174)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:166)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:95)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
        at io.ballerina.projects.Package.getCompilation(Package.java:135)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:64)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:163)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

**Affected Versions:**
slbeta2

**Related Issues:**
https://github.com/ballerina-platform/ballerina-lang/issues/29753

### LINKED ISSUE #33331: Enable test cases disabled by PR #33328
**Description:**
Enable test cases disabled by PR #33328

**TODO**
- [x] https://github.com/ballerina-platform/ballerina-lang/issues/32070
- [x] https://github.com/ballerina-platform/ballerina-lang/issues/32044

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by KRVPerera**
Fixes #32044
Fixes #33331
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/32617

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -470,6 +477,41 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
...
 
+    private void populateTypeToTypeDefMap(List<BLangNode> typeDefNodes) {
+        for (BLangNode typeDef : typeDefNodes) {
+            if (typeDef.getKind() == NodeKind.CLASS_DEFN) {
+                continue;
+            } else if (typeDef.getKind() == NodeKind.TYPE_DEFINITION) {
+                BLangTypeDefinition typeDefNode = (BLangTypeDefinition) typeDef;
+                BType type = typeDefNode.typeNode.getBType();
+                typeToTypeDef.put(type, typeDefNode);
+            }
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:488]: Can we not simply check if `typeDef.getKind() != NodeKind.TYPE_DEFITION` then continue and do the map insertion?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -470,6 +477,41 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
...
+        }
+    }
+
+    private void defineReferencedFieldsOfClassDef(BLangClassDefinition classDefinition, SymbolEnv pkgEnv) {
+        SymbolEnv typeDefEnv = classDefinition.typeDefEnv;
+        BObjectTypeSymbol tSymbol = (BObjectTypeSymbol) classDefinition.symbol;
+        BObjectType objType = (BObjectType) tSymbol.type;
+
+        // todo: check for class fields and object fields
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:511]:
```suggestion
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 3]
```diff
@@ -3489,6 +3530,34 @@ private void defineRecordTypeNode(BLangRecordTypeNode recordTypeNode, SymbolEnv
         symResolver.markParameterizedType(recordType, fieldTypes);
     }
 
+    private void resolveRestFieldType(BLangRecordTypeNode recordTypeNode, BRecordType recordType) {
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:3533]: Do we need this method anymore?

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:3533]: https://github.com/ballerina-platform/ballerina-lang/pull/35475/files#diff-fa94fe90a875c08eaf806f89f6ceb7eec7205acaf75c88a41e3f7103abfa5a52R3505 Seems it is not being used

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -3820,7 +3891,7 @@ public void defineFieldsOfClassDef(BLangClassDefinition classDefinition, SymbolE
         }
 
         // todo: check for class fields and object fields
-        defineReferencedClassFields(classDefinition, typeDefEnv, objType, false);
+        //defineReferencedClassFields(classDefinition, typeDefEnv, objType, false);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:3894]: Hope we can delete this now?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 5]
```diff
@@ -3451,17 +3490,19 @@ public void visit(BLangXMLAttribute bLangXMLAttribute) {
...
+    private void defineRecordTypeNode(BLangRecordTypeNode recordTypeNode) {
         BRecordType recordType = (BRecordType) recordTypeNode.symbol.type;
         recordTypeNode.setBType(recordType);
 
         // Define all the fields
-        resolveFields(recordType, recordTypeNode, env);
+        resolveFields(recordType, recordTypeNode);
+        resolveFieldsIncluded(recordType, recordTypeNode);
 
+//        resolveRestFieldType(recordTypeNode, recordType);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:3505]: delete this

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 6]
```diff
@@ -429,6 +433,9 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
 
         // Define type def fields (if any)
         defineFields(typeAndClassDefs, pkgEnv);
+//        List<BLangNode> orderedList = reorderTypeDefsAndClasses(typeAndClassDefs, pkgEnv);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:436]: delete this

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typedefs/TypeDefinitionsTest.java [hunk 7]
```diff
@@ -169,19 +169,19 @@ public void testBinaryExprAssignments() {
...
-//    public void testRecordTypeResolving() {
-//        JvmRunUtil.invoke(recordFieldRes, "testRecordTypeResolving");
-//    }
-//
-//    @Test
-//    public void testRecordTypeResolvingWithTypeInclusion() {
-//        JvmRunUtil.invoke(recordFieldRes, "testRecordTypeResolvingWithTypeInclusion");
-//    }
+    @Test
+    public void testRecordTypeResolving() {
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typedefs/TypeDefinitionsTest.java:173]: Shall we include a test, where we include a record type from a different module?

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/typedefs/TypeDefinitionsTest.java:173]: Done

**Review by gimantha — APPROVED** [review]
Shall we change the base to 2201.1.0-stage?

**Comment by KRVPerera** [comment]
> Shall we change the base to 2201.1.0-stage?

Created a PR for the staging branch. We can close this if it is clean.
https://github.com/ballerina-platform/ballerina-lang/pull/35884

**Comment by KRVPerera** [comment]
Full build without tests and locally tested few stdlib modules

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 8]
```diff
@@ -4801,17 +4865,40 @@ private MarkdownDocAttachment getMarkdownDocAttachment(BLangMarkdownDocumentatio
...
+            if (referredType.tag != TypeTags.RECORD) {
+                continue;
+            }
+            var fields = ((BStructureType) referredType).fields.values();
+            for (BField field : fields) {
+                BType type = field.type;
+                BLangTypeDefinition typeDefinition = typeToTypeDef.get(type);
+                if (typeDefinition != null && typeDefinition.typeNode != null &&
+                        typeDefinition.typeNode.getKind() == RECORD_TYPE) {
+                    defineReferencedFieldsOfRecordTypeDef(typeDefinition);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:4896]: Would you happen to remember why/if we needed this? There seems to be a codecov warning too. 

When the type here refers to a record type like

```ballerina
type Foo record {|
    int i;
|};
```

`type` is a `BTypeReferenceType`, so `typeDefinition` is always null.

This also leads to a SO error with something like

```ballerina
public type InnerErrorModel record {
    string code?;
    record {*InnerErrorModel;} innerError?;
};
``` 

https://github.com/ballerina-platform/ballerina-lang/issues/34346#issuecomment-1259001825

**Pull request merged**